internal/hcsoci/resources.go

```diff
@@ -33,3 +40,120 @@ func NormalizeMemorySize(ctx context.Context, cid string, requestedSizeMB uint64
 	}
 	return actualMB
 }
+
+// NormalizeUVMUpdateResourcesRequest accepts an unmarshalled [spec.UpdateTaskRequest.Resources]
+// and the accompanying annotations and returns the Resources struct with the appropriate fields
+// updated based on the VM resource annotations.
+func NormalizeUVMUpdateResourcesRequest(ctx context.Context, data interface{}, annots map[string]string) (interface{}, error) {
+	// calling this directly in uvm.Update() would cause an import cycle, so the best alternative is to
+	// call from containerd-shim-runhcs
+	return normalizeUpdateResourcesRequest(ctx, false, data, annots)
+}
+
+// NormalizeContainerUpdateResourcesRequest is similar to [NormalizeUVMUpdateResourcesRequest],
+// except it uses container resource annotations.
+func NormalizeContainerUpdateResourcesRequest(ctx context.Context, data interface{}, annots map[string]string) (interface{}, error) {
+	return normalizeUpdateResourcesRequest(ctx, true, data, annots)
+}
+
+func normalizeUpdateResourcesRequest(ctx context.Context, isContainer bool, data interface{}, annots map[string]string) (interface{}, error) {
+	var (
+		memorySizeInMB  = annotations.MemorySizeInMB
+		processorCount  = annotations.ProcessorCount
+		processorLimit  = annotations.ProcessorLimit
+		processorWeight = annotations.ProcessorWeight
+	)
+	if isContainer {
+		memorySizeInMB = annotations.ContainerMemorySizeInMB
+		processorCount = annotations.ContainerProcessorCount
+		processorLimit = annotations.ContainerProcessorLimit
+		processorWeight = annotations.ContainerProcessorWeight
+	}
```

Comment:
I'm not a huge fan of this but also not sure off the top of my head a better way of doing this :P